FiltersEngine.cpp

```diff
@@ -0,0 +1,201 @@
+
+    The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+
+    THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+*/
+#include "FiltersEngine.h"
+
+#include "Logger.h"
+
+#include <string>
+#include <iostream>
+#include <fstream>
+#include <sys/stat.h> /* for stat() */
+#include <dirent.h>
+#include <algorithm>
+#include <unistd.h>
+#include <limits.h>
+
+std::bitset<FILTER_BITSET_SIZE> FiltersEngine::AddFilter(ProcFilterSpec& pfs, std::string& outputName)
+{
+    std::bitset<FILTER_BITSET_SIZE> ret;
+
+    if (_filtersBitPosition.count(pfs) > 0) {
+        struct FiltersInfo info = _filtersBitPosition[pfs];
+        ret[info.bitPosition] = 1;
+        if (info.outputs.count(outputName) == 0) {
+            info.outputs.emplace(outputName);
+            _filtersBitPosition[pfs] = info;
+        }
+    } else {
+        struct FiltersInfo info;
+        info.bitPosition = _nextBitPosition;
+        info.outputs.emplace(outputName);
+        _filtersBitPosition[pfs] = info;
+        ret[_nextBitPosition] = 1;
+        _bitPositionSyscalls[_nextBitPosition] = pfs._syscalls;
+        _nextBitPosition++;
+    }
+
+    if (_outputs.count(outputName) == 0) {
+        _outputs.insert(outputName);
+    }
+
+    return ret;
+}
+
+std::bitset<FILTER_BITSET_SIZE> FiltersEngine::AddFilterList(std::vector<ProcFilterSpec>& pfsVec, std::string& outputName)
+{
+    std::bitset<FILTER_BITSET_SIZE> ret;
+
+    for (auto pfs : pfsVec) {
+        ret |= AddFilter(pfs, outputName);
+    }
+
+    return ret;
+}
+
+bool FiltersEngine::ProcessMatchFilter(std::shared_ptr<ProcessTreeItem> process, ProcFilterSpec& pfs, unsigned int height)
+{
+    if (pfs._depth != -1 && pfs._depth < height) {
+        return false;
+    }
+
+    if (pfs._match_mask & PFS_MATCH_UID) {
+        if (pfs._uid != process->_uid) {
+            return false;
+        }
+    }
+
+    if (pfs._match_mask & PFS_MATCH_GID) {
+        if (pfs._gid != process->_gid) {
+            return false;
+        }
+    }
+
+    if (pfs._match_mask & PFS_MATCH_EXE_EQUALS) {
+        if (pfs._exeMatchValue != process->_exe) {
```

Comment:
Guess casing is more consistent in the UNIX world than Windows so just a question really whether we should be doing a case -insensitive match for these strings?

Reply:
For executables case is specific. For arguments it could be slightly more lenient but generally it is pretty specific. Maybe we could consider adding a case-insensitivity flag of some sort?